ENH: Implement sliding window

zklaus opened the pull request:
This is an updated version of #10771. It implements the sliding window function discussed in #7753 according to the slightly changed API as suggested by @eric-wieser in https://github.com/numpy/numpy/pull/10771#issuecomment-524715356.

# The case for a sliding window function
A sliding window function is a longstanding open demand. Issue #7753 has been open since 2016 and the discussion about this feature has come up several times since then. Additionally, several downstream libraries, such as dask, scikit-image, pandas, etc. need this functionality; some of them already have their own version, see eg [`skimage.utils.view_as_windows()`](https://scikit-image.org/docs/stable/api/skimage.util.html#skimage.util.view_as_windows) and [`skimage.utils.view_as_blocks()`](https://scikit-image.org/docs/stable/api/skimage.util.html#skimage.util.view_as_blocks). However, due to the general nature of this functionality it is believed to have a place in numpy itself.

The one counter-argument to the inclusion into numpy that has been made is that in many cases where this function could be used, more efficient algorithms are available (see e.g. https://github.com/numpy/numpy/issues/7753#issuecomment-227430211, https://github.com/numpy/numpy/issues/7753#issuecomment-597804716). Nevertheless, the consensus reached in #7753 seems to be that, while often more efficient algorithms would be available, people rarely implement these problem-specific optimizations, but rather opt for a simpler approach, either based on explicit loops or on one of several sliding window implementations that are "in the wild" (see e.g. https://gist.github.com/teoliphant/96eb779a16bd038e374f2703da62f06d). In both of these cases having a well-documented, standardized function in numpy will enhance readability, reduce code duplication, and potentially improve performance.

In summary, a sliding window function is a worthwhile addition to numpy and will close a longstanding open issue.

# The specific API
The API as proposed in this PR is the result of a discussion in the ancestral PR #10771. It is similar to the one used in skimage, but adds the axis keyword to allow flexible windows over any combination of axes, including repeated axes. Several use-cases are demonstrated [here](https://github.com/numpy/numpy/pull/10771#issuecomment-524715356); a possible implementation of the skimage functions based on this API is given [here](https://github.com/numpy/numpy/pull/10771#issuecomment-601121403).

Closes #7753.

Commit:
implement sliding_window_view #7753
Test cases are shown in the issue page.

Commit:
Add Example Cases

Commit:
Add step_size and N-dim support

Commit:
Add shape and step_size check. Remove warning.

Commit:
Remove shape default
Add step_size default's description.

Commit:
Give proper parameter name 'step'

Commit:
fix a parameter description mistake

Commit:
implement test function for sliding_window_view()

Commit:
implement test function for sliding_window_view()

Commit:
Fix according to @eric-wieser comments

Commit:
Change arange to ogrid in Examples

Commit:
remove np.squeeze on return line

fnjn committed:
Clarify document to avoid parameter confusion.

Commit:
add `writable` and more explanation in docs

Commit:
resolve a write conflit

Commit:
fixes according to @seberg review

Commit:
resolve write hazard

Commit:
remove outdated docs.

Commit:
change referring according to @mattip.
change 'writeable' to 'readonly' as @seberg suggest.
remove 'step' as @eric-wieser request

Commit:
fix test minor error

eric-wieser committed:
DOC: Grammar fixes

eric-wieser committed:
STY: Add missing line break required by PEP8

fanjin-z committed:
+ Change readonly parameter to writeable.
+ Update writeable description.
+ Fix a few parameter checks.
+ Other minor improvements.

Commit:
Move to new api as proposed by @eric-wieser

- Change api to follow suggestion by Eric Wieser in
  https://github.com/numpy/numpy/pull/10771#issuecomment-524715356
- Update docstring
- Add more tests

Commit:
Improve documentation

Commit:
Add sliding_window_view to documentation index

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    x : array_like
+        Array to create the sliding window view from.
+
+    shape : int or tuple of int
+        The shape of the window. If `axis` is not present, must have same
+        length as the number of input array dimensions.
+
+    axis : int or tuple of int, optional
+        If present, `shape[i]` will refer to the axis `axis[i]` of `x`,
+        otherwise `shape[i]` will refer to axis `i` of `x`.
```
```suggestion
    shape : int or tuple of int
        The shape of the window. If `axis` is not present, must have same
        length as the number of input array dimensions.
        Single integers `i` are treated as if they were the tuple `(i,)`.

    axis : int or tuple of int, optional
        If present, `shape[i]` will refer to the axis `axis[i]` of `x`,
        otherwise `shape[i]` will refer to axis `i` of `x`.
        Single integers `i` are treated as if they were the tuple `(i,)`.
```

zklaus replied:
Good clarification, thanks!

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
     return view
 
 
+def sliding_window_view(x, shape, axis=None, subok=False, writeable=False):
```
I think we should make these kwarg-only, either as
```suggestion
def sliding_window_view(x, shape, axis=None, *, subok=False, writeable=False):
```
or
```suggestion
def sliding_window_view(x, shape, *. axis=None, subok=False, writeable=False):
```
depending on whether we want to force the caller to write the more explicit `sliding_window_view(x, s, axis=a)` instead of `sliding_window_view(x, s, a)`.

zklaus replied:
Good idea! I like more the former since I expect that axis will almost always be provided.

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    shape_array = np.array(shape)
+    if np.any(shape_array < 0):
+        raise ValueError('`shape` cannot contain negative value')
```
```suggestion
        raise ValueError('`shape` cannot contain negative values')
```

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        raise ValueError(
+            'window shape cannot be larger than input array shape')
+
+    if axis is None:
+        axis = np.arange(x.ndim)
+    axis = tuple(axis) if np.iterable(axis) else (axis,)
```
```suggestion
    if axis is None:
        axis = tuple(range(x.ndim))
    else:
        axis = normalize_axis_tuple(axis, x.ndim, allow_duplicate=True)
```
which ensures that `AxisError` is thrown for invalid axes, rather than an `IndexError`.

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        axis = np.arange(x.ndim)
+    axis = tuple(axis) if np.iterable(axis) else (axis,)
+
+    w_ndim = len(shape)
+    assert w_ndim == len(axis)
```
```suggestion
    if len(shape) != len(axis):
        raise ValueError(f"Must provide matching length shape and axes; got {len(shape)} shape elements and {len(axes)} axes elements)
```

It might be nice to have a separate error for `axis=None`, to explain to the user that if `axis=None`, all the dimensions are expected.

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    shape_array = np.array(shape)
+    if np.any(shape_array < 0):
+        raise ValueError('`shape` cannot contain negative value')
+    if np.any(x.shape < shape_array):
```
I suspect this logic is wrong since axes can be repeated.... Let's see if tests can tell us.

zklaus replied:
You are right. This is a remnant from the version without axis keyword. Will address in new commit.

eric-wieser reviewed (commented):
Thanks for picking this back up.

Would you be able to add some tests for the error cases? In case you're not familiar with pytest, you can do that with
```python
with pytest.raises(ValueError, match="cannot be larger"):
    sliding_window_view(...)
```

BvB93 commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    """
+    shape = tuple(shape) if np.iterable(shape) else (shape,)
+    # first convert input to array, possibly keeping subclass
+    x = np.array(x, copy=False, subok=subok)
```
```suggestion
    x = np.asanyarray(x)
```

zklaus replied:
This was briefly discussed in the previous PR https://github.com/numpy/numpy/pull/10771#discussion_r317427162. There, @eric-wieser suggested that this was done on purpose and to conform with `as_strided`. That's why I kept it, but I have to admit that I don't understand the issue in detail.

@eric-wieser, would you care to way in?

BvB93 replied:
IMO this sounds more like an argument for changing `as_strided`. 
Still, as this is quite minor I don't have particularly strong feelings about either option; feel free to ignore.

eric-wieser replied:
I think we should leave this as is for now. `_broadcast_to` also uses the same pattern. @BvB93, if you feel strongly about this feel welcome to open a follow-up PR that changes all the conversions in this file.

BvB93 commented:
The module-level `__all__` should probably be updated as well.

Commit:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Commit:
Fix window shape check

Commit:
Add `sliding_window_view` to __all__

BvB93 commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,141 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
     return view
 
 
+def sliding_window_view(x, shape, axis=None, *, subok=False, writeable=False):
```
```suggestion
@set_module('numpy')
def sliding_window_view(x, shape, axis=None, *, subok=False, writeable=False):
```

This should fix the CI failure.
Note that `set_module` will also have to be imported from `numpy.core.overrides`.

eric-wieser replied:
Or perhaps we should add the full array_function dispatching

zklaus replied:
I'm sorry, I don't know what that means. Could you please clarify? In the meantime, I'll get to work on those error tests...

zklaus replied:
Ok, I read (parts of) NEP-18 and took a stab at adding the array_function dispatching. One thing that I am still unsure about is the usage of defaults for arguments. Please have a look and let me know what you think.

Commit:
Add tests for  error cases

Commit:
Add array_function dispatching

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,152 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+
+def _sliding_window_view_dispatcher(x, shape, axis=None, *,
+                                    subok=False, writeable=False):
```
Dispatcher defaults should be None
```suggestion
                                    subok=None, writeable=None):
```

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,152 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+           [[[20, 21, 22],
+             [21, 22, 23]]]])
+    """
+    return _sliding_window_view(x, shape, axis,
```
I'm a bit puzzled by this underscore version existing, but it seems to match the rest of the file.

zklaus replied:
So I'm not the only one :smile: Yeah, I only followed the rest of the file's pattern.

eric-wieser replied:
Ah; the rest of the file does it to avoid dispatching twice. There's no need to do that here, because you have no other callers of `_sliding_window_view`.

So can you remove `_sliding_window_view` and paste its contents here?

zklaus replied:
Ah yes, that makes sense.

Commit:
Change dispatcher argument defaults to None

Commit:
Simplify array function dispatching

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -333,6 +336,69 @@ def test_as_strided():
     assert_array_equal([r] * 3, a_view)
 
+
+def test_sliding_window_view():
```
I think it would be nice to split this into smaller tests. One thing you can do is wrap the tests in a class,

```
class TestSlidingWindowView:
    def test_1d(self): ...

    def test_2d(self):

    def test_errors(self):
```
Shorter tests are easier to follow, as you don't have to keep track of local variables for as long.

Up to you how you want to split these - every time you start with a new array would be a reasonable choice.

BvB93 commented:
I assume this will need a release note? 
Instructions and a few release note examples are contained within [doc/release/upcoming_changes](https://github.com/numpy/numpy/tree/master/doc/release/upcoming_changes).

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,147 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    Examples
+    --------
+    >>> x = np.arange(5)
+    >>> sliding_window_view(x, 2)
+    array([[0, 1],
+           [1, 2],
+           [2, 3],
+           [3, 4]])
```
In my opinion this is slightly easier to understand with a bigger array,
```suggestion
    >>> x = np.arange(6)
    >>> sliding_window_view(x, 3)
    array([[0, 1, 2],
           [1, 2, 3],
           [2, 3, 4],
           [3, 4, 5]])
```

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,147 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+           [1, 2],
+           [2, 3],
+           [3, 4]])
+
```
I think it would be nice to show this use-case too:
```suggestion
    Combining with stepped slicing, this can be used to take sliding views which skip elements:

    >>> x = np.arange(7)
    >>> sliding_window_view(x, 5)[:, ::2]
    array([[0, 2, 4],
           [1, 3, 5],
           [2, 4, 6]])

    or views which move by multiple elements
     
    >>> x = np.arange(7)
    >>> sliding_window_view(x, 5)[::2, :]
    array([[0, 1, 2],
           [2, 3, 4],
           [4, 5, 6]])

```

BvB93 commented:
There are currently a few CI errors caused by the doctests, 
the latter throwing a `NameError` when encountering `sliding_window_view`.

Replacing all references to `sliding_window_view` with `np.sliding_window_view` in the code blocks should fix the issue.

```
File "build/testenv/lib/python3.6/site-packages/numpy/__init__.py", line 95, in sliding_window_view
Failed example:
    sliding_window_view(x, (2, 3), (1, 1))
Exception raised:
    Traceback (most recent call last):
      File "/Users/runner/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/doctest.py", line 1330, in __run
        compileflags, 1), test.globs)
      File "<doctest sliding_window_view[8]>", line 1, in <module>
        sliding_window_view(x, (2, 3), (1, 1))
    NameError: name 'sliding_window_view' is not defined
```

Commit:
Added "np." prefix to doctests

Commit:
Split tests

Commit:
Improved docstring

Commit:
Add release note

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -203,21 +203,36 @@ def sliding_window_view(x, shape, axis=None, *, subok=False, writeable=False):
            [[[20, 21, 22],
              [21, 22, 23]]]])
+
+    Combining with stepped slicing (`::step`), this can be used to take sliding
+    views which skip elements:
+    >>> x = np.arange(7)
+    >>> np.sliding_window_view(x, 5)[:, ::2]
+    array([[0, 2, 4],
+           [1, 3, 5],
+           [2, 4, 6]])
+    or views which move by multiple elements
```
I think this need blank lines
```suggestion
    Combining with stepped slicing (`::step`), this can be used to take sliding
    views which skip elements:

    >>> x = np.arange(7)
    >>> np.sliding_window_view(x, 5)[:, ::2]
    array([[0, 2, 4],
           [1, 3, 5],
           [2, 4, 6]])

    or views which move by multiple elements
```

zklaus replied:
You are right.

Commit:
Fix docstring formatting

zklaus commented:
SEGFAULT in test_umath? That doesn't feel like it comes from this PR, right?

BvB93 commented:
> SEGFAULT in test_umath? That doesn't feel like it comes from this PR, right?

Let's assume for know that this is a fluke; restarting the test will probably fix it.

More importantly, there is still a doctest-related error:
```
numpy.lib.sliding_window_view
-----------------------------

File "build/testenv/lib/python3.6/site-packages/numpy/__init__.py", line 115, in sliding_window_view
Failed example:
    np.sliding_window_view(x, 5)[::2, :]
Expected:
    array([[0, 1, 2],
           [2, 3, 4],
           [4, 5, 6]])
Got:
    array([[0, 1, 2, 3, 4],
           [2, 3, 4, 5, 6]])


ERROR:  failed checking numpy.lib

```

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,164 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    or views which move by multiple elements
+
+    >>> x = np.arange(7)
+    >>> np.sliding_window_view(x, 5)[::2, :]
```
```suggestion
    >>> np.sliding_window_view(x, 3)[::2, :]
```
should fix the doctest

eric-wieser commented on doc/source/reference/routines.indexing.rst:
```diff
@@ -43,6 +43,7 @@ Indexing-like operations
    diagonal
    select
    lib.stride_tricks.as_strided
+   lib.stride_tricks.sliding_window_view
```
```suggestion
   sliding_window_view
```
Note that `as_strided` is deliberately _not_ available as `np.as_strided` as it is not present in `__all__`.

I think it's fine that you have `sliding_window_view` in `__all__` in this PR, but you therefore shouldn't have the namespacing here.

zklaus replied:
Agreed. I added this before we put the array dispatch in place and forgot to update.

I am then also moving it in front of `as_strided` in this list, correct?

Commit:
Fix doctest

Commit:
Remove namespacing in documentation indexing

shoyer commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,164 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    Returns
+    -------
+    view : ndarray
+        Sliding window view of the array.
```
Two question here:

- Where do the sliding window dimensions get inserted -- at the start or at the end?
- Can you specify what the shape of the result will be, in terms of the original shape of the array?

zklaus replied:
Good points! Will update docstring. :+1:

zklaus replied:
I tried to clarify the docstring. As a result, I also renamed the `shape` argument to `window_shape`. I am not 100% pleased with the results of my efforts but didn't manage to come up with a better formulation.

Suggestions are very welcome!

shoyer commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,164 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    axis : int or tuple of int, optional
+        If present, `shape[i]` will refer to the axis `axis[i]` of `x`,
+        otherwise `shape[i]` will refer to axis `i` of `x`.
```
I wonder if it makes more sense for the default to be "all axes". Has this been considered?

zklaus replied:
I'm not sure I completely understand what you mean. In a sense, the default already is "all axes" in so far as `axis=None` or not present is equivalent to `axis=tuple(range(x.ndim))` and means you need to specify as many entries in `shape`.

Perhaps your suggestion is to change the meaning of single integers in `shape` from `(i,)` to `(i,)*len(axis)`?

eric-wieser replied:
Another option that @shoyer might be suggesting that `axis=-1` or `axis=0` is a sensible default. This might be the type of thing where the best thing to do is describe the possible choices at the top of this issue, and ultimately let the mailing list decide.

shoyer replied:
I misinterpreted the docstring. I was trying to suggest exactly what is already implemented!

I would suggest revising this to something like:

> Axis or axes along which the sliding window is applied. By default, the sliding window is applied to all axes.

And then for window_shape:

> Size of window over each axis that takes part in the sliding window...

zklaus replied:
Ok, seems the implementation is agreed; will work on the docstring.

zklaus replied:
Clarified docstring.

Commit:
Improve docstring

zklaus commented:
While trying to improve the documentation for the return value, I decided to change the name of the `shape` argument to `window_shape`. Is that a good idea? Do you oppose it?

charris commented:
Close/reopen

shoyer commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,190 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        required by the size of the sliding window. That is,
+        `view.shape = x_shape_trimmed + window_shape`, where `x_shape_trimmed` is
+        `x.shape` with every entry reduced by the size of the window in the
+        respective axis minus one.
```
Can you try just giving the formula here? From reading the text, it isn't clear whether "minus one" refers to the size of the window being subtracted or the final result. E.g., is it "input_size - (window_size - 1)" or "input_size - window_size - 1"?

eric-wieser replied:
The former.

zklaus replied:
Clarified docstring.

Commit:
Improved docstring

shoyer commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,206 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        window size in the corresponding window axis, then the respective entry
+        of `x_shape_trimmed` will be `input_size - (window_size - 1)`.
+
+        If the same axis is used several times, there will be several
+        associated window sizes. In this case the reduction proceeds in exactly
+        the same way as
+        `input_size - (window_size_1 - 1) - (window_size_2 - 1) - ...`.
+
+        For a better understanding of this, also refer to the examples below.
```
Deleting this paragraph? Looking at examples is recommended for understanding pretty much every function in NumPy :)

shoyer commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,206 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    Returns
+    -------
+    view : ndarray
+        Sliding window view of the array. The sliding window dimensions are
+        inserted at the end, and the original dimensions are trimmed as
+        required by the size of the sliding window.
+
+        That is, `view.shape = x_shape_trimmed + window_shape`, where
+        `x_shape_trimmed` is `x.shape` with every entry reduced as follows.
+
+        Let `input_size` be the size of the input array in an axis
+        participating in the sliding window, and let `window_size` be the
+        window size in the corresponding window axis, then the respective entry
+        of `x_shape_trimmed` will be `input_size - (window_size - 1)`.
```
This is a little wordy now :).

Maybe we could just give the formula, i.e.,
```suggestion
        Sliding window view of the array. The sliding window dimensions are
        inserted at the end, and the original dimensions are trimmed as
        required by the size of the sliding window. That is,
        `view.shape = x_shape_trimmed + window_shape` where
        `x_shape_trimmed[i] = x.shape[i] - (window_shape[i] - 1)`.
```

zklaus replied:
I completely agree that it is more wordy than desirable. The problem is that your suggested formula is not correct and while I was trying to come up with a correct formula, I ended up with things that I did not find more enlightening than the text. The correct formula would be `x_shape_trimmed[i] = x.shape[i] if i not in axis` and otherwise
```
x_shape_trimmed[axis[j]] = x.shape[axis[j]] - (window_shape[j] - 1)
```
except for the doubly used axis, in which case you need to subtract the contribution from all window axis (avoid double counting).

eric-wieser replied:
Right - the challenge here is to describe exactly what these lines of code do:

https://github.com/numpy/numpy/blob/97315fc86e7e8dc49aa659eb86428c915bf02375/numpy/lib/stride_tricks.py#L304-L311

IMO we maybe want to lean more in a "they do the obvious thing" direction - what's important is that:
* The new `window_shape` dimensions are added at the end
* The number of initial dimensions are unchanged, but the length of each dimensions will be adjusted to account for the windowing

shoyer commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,206 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        window size in the corresponding window axis, then the respective entry
+        of `x_shape_trimmed` will be `input_size - (window_size - 1)`.
+
+        If the same axis is used several times, there will be several
+        associated window sizes. In this case the reduction proceeds in exactly
+        the same way as
+        `input_size - (window_size_1 - 1) - (window_size_2 - 1) - ...`.
```
Move this part into the "Notes" section below? Repeated axes are a rarely used edge case.

zklaus replied:
Good idea! Let's see how the other issues pan out and then move/reformulate accordingly.

zklaus commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,206 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        required by the size of the sliding window.
+
+        That is, `view.shape = x_shape_trimmed + window_shape`, where
+        `x_shape_trimmed` is `x.shape` with every entry reduced as follows.
+
+        Let `input_size` be the size of the input array in an axis
+        participating in the sliding window, and let `window_size` be the
+        window size in the corresponding window axis, then the respective entry
+        of `x_shape_trimmed` will be `input_size - (window_size - 1)`.
+
+        If the same axis is used several times, there will be several
+        associated window sizes. In this case the reduction proceeds in exactly
+        the same way as
+        `input_size - (window_size_1 - 1) - (window_size_2 - 1) - ...`.
+
+        For a better understanding of this, also refer to the examples below.
```
```suggestion
        `x_shape_trimmed` is `x.shape` with every entry reduced by one less than the corresponding window size.
```

zklaus replied:
How about this?

shoyer replied:
Looks great!

Commit:
Simplified docstring

zklaus commented:
Looks like all comments are addressed. @shoyer, @eric-wieser, @BvB93 are you guys happy with the result? What's the next step?

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,196 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        inserted at the end, and the original dimensions are trimmed as
+        required by the size of the sliding window.
+
+        That is, `view.shape = x_shape_trimmed + window_shape`, where
+        `x_shape_trimmed` is `x.shape` with every entry reduced by one less
```
```suggestion
        That is, `view.shape = x_shape_trimmed + window_shape`, where
        ``x_shape_trimmed`` is ``x.shape`` with every entry reduced by one less
```

zklaus replied:
Why do you suggest the use of double backticks here? Everything I read [here](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) seems to suggest single backticks in this case.

eric-wieser replied:
From that page:
```
Use *italics*, **bold** and ``monospace`` if needed in any explanations (but not for variable names and doctest code or multi-line code). Variable, module, function, and class names should be written between single back-ticks (`numpy`).
```
I'd argue that `x.shape` is not a name, so does not fall into the second sentence.

Primarily though, my argument is just that these bits would be easier to read in the docs (which I link above) if they were monospaced

zklaus replied:
Ah, yes, agree :+1:

eric-wieser reviewed (commented):
Rendered docs are here: https://16134-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.sliding_window_view.html

I think we should probably ping the mailing list here, to check this does what everyone needs it to. It might be worth assembling a summary of the discussion in the other issue first, and pasting it in the description in this PR.

Commit:
Improve docstring to make pseudo code stand out

zklaus commented:
I updated the description at the top of this PR to summarize the discussion around this issue. Ready to ping the mailing list?

bashtage commented:
Any chance of getting a non-trivial doc page (or addition to an existing one showing how this can be used?

bashtage commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,196 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    ----------
+    x : array_like
+        Array to create the sliding window view from.
+
```
Are spaced between variables common? I thought normally these should be packed with only indentation indicating a variable after a description.

bashtage commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,196 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    Examples
+    --------
+    >>> x = np.arange(6)
```
Should there be an example showing how the sliding window view can be used say to compute a sliding window mean similar to `pd.rolling`

charris replied:
I was wondering if it could be used for an IIR filter, although I think SciPy already has a function for that.

zklaus replied:
I am not knowledgable enough to construct an IIR example, but I will include a simple moving average.

zklaus replied:
Thanks for your comments!

I added a simple moving average example. If you consider this not sufficient in terms of application examples, I am open for other suggestions. Otherwise, I will not consider this a blocker for merging.

bashtage commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,196 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    """
+    Create a sliding window view into the array with the given window shape.
+
+    Creates a sliding window view of the N dimensional array with the given
```
Perhaps note that these are called rolling windows in some domains somewhere in the docstring to improve discoverability.

eric-wieser commented:
@bashtage, are you asking for more than the page I link in https://github.com/numpy/numpy/pull/17394#pullrequestreview-503740766?

bashtage reviewed (commented):
Some minor suggestions that might help users discover and use this view.

bashtage commented:
> @bashtage, are you asking for more than the page I link in [#17394 (review)](https://github.com/numpy/numpy/pull/17394#pullrequestreview-503740766)?

"Asking" is maybe a but much.  Something that shows the power of the view in application -- for example moving window statistics or something from anyother domain where this is useful -- rather than just how it works as a view, would help end-users discover it more easily IMO.

bashtage commented:
Another question: How does the performance of this view compare to using a Cython/numba loop for sliding window calculations?

Commit:
Improve docstring

Commit:
Add simple application example

eric-wieser commented:
@bashtage

> How does the performance of this view compare to using a Cython/numba loop for sliding window calculations?

The cost of making the view is `O(1)`, so it shouldn't matter. What matters is the algorithm for computing over the window - for instance, a moving sum of length `n` is faster to compute as `c = np.cumsum(x); return c[n:] - c[:-n]` than as `sliding_window(x, (n,)).sum(axis=-1)`

eric-wieser commented on doc/release/upcoming_changes/17394.new_function.rst:
```diff
@@ -0,0 +1,5 @@
+``sliding_window_view`` provides a sliding window view for numpy arrays
+-----------------------------------------------------------------------
+``np.sliding_window_view`` constructs views on numpy arrays that offer a sliding
```
```suggestion
`numpy.sliding_window_view` constructs views on numpy arrays that offer a sliding
```
so that this makes a link (`np` does not work in links)

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,211 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+def sliding_window_view(x, window_shape, axis=None, *,
+                        subok=False, writeable=False):
+    """
+    Create a sliding window view into the array with the given window shape.
+
+    Creates a sliding window view (also known as rolling or moving window) of
+    the N dimensional array with the given window shape. The window slides
```
It seems a bit odd to have this sentence repeated, the second time in slightly more detail. Not a blocker against merging, but maybe someone on the doc team can offer a better phrasing.

Commit:
Correct release note

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,211 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    See Also
+    --------
+    lib.stride_tricks.as_strided: Create a view into the array with the
+        given shape and strides.
```
```suggestion
    lib.stride_tricks.as_strided: A lower-level and less safe routine for
        creating arbitrary views from custom shape and strides.
```

eric-wieser replied:
Also, you should add a link in the other direction too

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +112,211 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    >>> moving_average = v.mean(axis=-1)
+    >>> moving_average
+    array([1., 2., 3., 4.])
+    """
```
I wonder if we should expand on the performance note here too:

```suggestion
    
    Note that this also provides an example of a case when `sliding_window_view` is not the fastest algorithm.
    The above needed 8 additions, but we can do it in 6 additions and 1 subtraction with `cumsum`:
    
    >>> cx = np.cumsum(x)
    >>> moving_average = (cx[3:] - cx[:-3]) / 3
    >>> moving_average
    array([1., 2., 3., 4.])
    """
```

This does overlap with the `notes` section...

zklaus replied:
Well, this also shows the difficulty with other algorithms: Your calculation misses the first window and produces `[2, 3, 4]` only. Is this fixable?

eric-wieser replied:
Ah, I always forget that complexity. Yes, either `cx` or `x` needs a zero prepending.

zklaus replied:
So, just to clarify. Let `n = len(x)` the length of the vector and `w = 3` the window size. Then the moving window approach needs `(n - (w-1))*(w-1) = (w-1)n - (w-1)^2 = 8` additions. The cumsum approach needs `n - 1` additions for the cumsum and `n - (w-1))` additions for the difference for a total of `n-1 + n - (w-1) = 2n - w = 9` additions, right? Granted, the larger the window, the more the moving window approach takes, but perhaps for moderate data and window sizes the example is not so convincing? Unless I got my numbers wrong, of course.

zklaus replied:
Both are O(n), for most window sizes the factor is better for cumsum than for rolling window, but the window approach is more parallelizable since every calculation is local.

I wonder how much of this discussion really should be in the docstring of the function.

seberg replied:
Both are O(n) in the length, but a [summed area table](https://en.wikipedia.org/wiki/Summed-area_table) is O(1) in the *window* size.  We probably do not need to give an exact calculation, but a rough estimation that this approach scales with the window size, while a summed area table does not (and thus is magnitudes faster for larger windows) is important IMO, since it should be quite typical that this is a nice quick solution, but rarely a particular performant one.

eric-wieser replied:
Whoops, I miscounted the subtraction

> Then the moving window approach needs `(n - (w-1))*(w-1) = (w-1)n - (w-1)^2 = 8` additions. The cumsum approach needs `n - 1` additions for the cumsum and `n - (w-1))` additions for the difference for a total of `n-1 + n - (w-1) = 2n - w = 9` additions, right? 

That sounds correct to me - as @seberg  points out, the key property is that the former is quadratic in `w`.

zklaus replied:
Hm. I agree with the fact, of course, but the wording is a bit strong for me. After all, the scaling is indeed linear in the operations, but flops are (essentially) for free these days; memory and communication is not. A summed area table can be calculated in one pass, but only serially and it will require either destruction of the original data for an in-place summation or double the memory requirement. So over all, for modern problems in a heterogeneous, multi-level memory, big data problem, the balance is not quite as obvious to me.

shoyer replied:
I suspect `cumsum` would also be prone to overflow.

That said, there are other methods for rolling window sum. Pandas and bottleneck use an approach that also runs in O(1) with respect to the window size, where a single element is added and subtracted from the rolling sum in each step.

seberg replied:
I am happy to tune it down, but I think the statement should include something to the fact that this can easily be hundreds of times slower than a proper solution and that such a proper solution is sometimes readily available! I guess the "100 times" may sound a bit too strong, but users may well be unaware of the big O notation and algorithmic complexity.

So, I think we should cover:

* For large windows, this can be orders of magnitude slower than an optimal approach.
* There are often much better solutions (e.g. in scipy), users should consider looking for them.
* Yes, for small windows there is probably no issue, and even for larger windows: This is will often be a very practical, sound solution (at least as long as the window sizes is fixed/limited)

For the summed are table, I don't know, maybe we should not mention it (it is a nice, simple example, but chances are fftconvolve and not building the full table are just better solutions).

@shoyer good point about bottleneck, we could also mention that here (although a lot of the tools are probably also available in `scipy.ndimage`, albeit probably slower for certain tasks.

zklaus replied:
Good input, thanks guys! I added a note in line (I believe) with @seberg's suggestion as well as links to scipy.ndimage and bottleneck (thanks for that, @shoyer).

What do you think about this formulation?

zklaus replied:
For ease of review, the proposed formulation is this (see [here](https://16504-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.sliding_window_view.html#numpy.sliding_window_view) for full rendered documentation):

At the moving average example I added a comment stating:
> Note that a sliding windows approach is often **not** optimal (see Notes).

And in the notes section the following:
> For some cases there may be more efficient approaches to calculate
transformations across multi-dimensional arrays, for instance
`scipy.signal.fftconvolve`, where combining the iterating step with the
calculation itself while storing partial results can result in significant
speedups.
>
> As a rough estimate, a sliding window approach with an input size of `N`
and a window size of `W` will scale as `O(N*W)` where frequently a special
algorithm can achieve `O(N)`. That means that the sliding window variant
for a window size of 100 can be a 100 times slower than a more specialized
version.
> For many common problems these better, specialized solutions already exist
in libraries such as
`scipy.ndimage <https://docs.scipy.org/doc/scipy/reference/ndimage.html>`_
or `bottleneck <https://github.com/pydata/bottleneck>`_.
>
> Nevertheless, for small window sizes, when no custom algorithm exists, or
as a prototyping and developing tool, this function can be a good solution.

zklaus replied:
@seberg, sorry, can I push this? It seems we are almost there. If you feel this needs stronger language, I am fine with that too. That is certainly not the hill I want (this PR) to die on.

bashtage commented:
What about a sliding covariance? Is the nontrivial enough that it can't be easily done with a cumsum.

eric-wieser commented:
Chances are if you understand the implementation of your single-pass algorithm, you can always come up with a faster implementation. In the case of covariance, you can compute the sliding means and sliding means of the products using a `cumsum` based approach.

`sliding_window` is still a very useful tool for prototyping techniques, but at the point you start worrying about performance, the problem is your algorithm not the implementation of `sliding_window`.

Commit:
Improve link with as_strides

Commit:
Add note about performance

Commit:
Tidy up main doc string

seberg commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +115,225 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    For many common problems these better, specialized solutions already exist
+    in libraries such as
+    `scipy.ndimage <https://docs.scipy.org/doc/scipy/reference/ndimage.html>`_
+    or `bottleneck <https://github.com/pydata/bottleneck>`_.
```
I like `window_shape`. I am OK with the current text. Personally, I would prefer it lead slightly stronger.  Not sure, maybe, leading with.

For many applications using a sliding view can be convenient but potentially very slow. Often specialized solutions solution exists, for example:

* `scipy.signal.fftconvolve`
* Filtering functions in `scipy.ndimage`
* Moving window functions provided by `bottleneck <https://bottleneck.readthedocs.io/en/latest/intro.html>`_

...

seberg replied:
But, at some point, we should probably call it a day and leave it to a future PR.

seberg commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -394,6 +396,84 @@ def test_as_strided():
+                match='window shape cannot be larger than input array'):
+            sliding_window_view(arr, (5, 5))
+
+
```
Could you add a test for the writeable argument? (I know it is trivially inherited, but would be nice to ensure API can't change accidentally). `subok` might be good as well.

zklaus replied:
Good thing you requested this: Actually the arguments were not passed on correctly. Writeable test is in, working on subok now.

zklaus replied:
Subok test is in as well.

Commit:
Make language on performance warning stronger

Commit:
Bugfix: pass subok and writeable to as_strided

Commit:
Add writeable test

Commit:
Add subok test

zklaus commented:
All outstanding comments are addressed. While doing the `subok` test I noticed that this is not working properly with `MaskedArray`, but I suggest leaving this to another PR that adds the masked code via dispatching. If that's ok, I'll add an issue to track it and get to work there asap.

Ok?

Otherwise, I think this is ready for merge.

seberg commented:
I think if you do the MA test, it should have a comment saying that the result is nonsense. Mayb it is better to just use
```
class MyArray(np.ndarray):
    pass

arr = np.arange(5).view(MyArr)
```

Commit:
Change subok test to use custom array subclass instead of unsupported MaskedArray

zklaus commented:
The Travis CI takes forever to queue tests now, but the tests run locally and worked on azure as well.

I changed the subok test as requested (makes sense). Anything else?

zklaus commented:
Addressed the last comment on subok test, rebased on current master, made sure all checks pass. The previous travis ci problem was most likely due to the move to travis-ci.com, now everything seems ok.

@seberg, good to go?

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +115,226 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    Also known as rolling or moving window, the window slides across all
+    dimensions of the array and extracts subsets of the array at all window
+    positions.
```
```suggestion
    positions.
    
    .. versionadded:: 1.20.0
```

Commit:
Add version added information

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

zklaus commented:
@seberg, @shoyer, @eric-wieser, is there anything left for me to do here? How can we move forward?

seberg commented:
Thanks @zklaus, I am sure we can always find some nitpicks, but it seems all good to me.  If there are any followups we can open a new PR. Thanks everyone who reviewed this!

shoyer commented:
Thank you @zklaus!

zklaus commented:
Cheers guys! Nice working with you :smiley:
